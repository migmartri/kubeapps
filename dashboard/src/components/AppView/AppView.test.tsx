import { shallow } from "enzyme";
import context from "jest-context";
import { safeDump as yamlSafeDump } from "js-yaml";
import * as React from "react";

import { hapi } from "../../shared/hapi/release";
import sharedSpecs from "../../shared/specs";
import { ForbiddenError, IResource, NotFoundError } from "../../shared/types";
import DeploymentStatus from "../DeploymentStatus";
import { NotFoundErrorAlert, PermissionsErrorAlert } from "../ErrorAlert";
import AppControls from "./AppControls";
import AppDetails from "./AppDetails";
import AppNotes from "./AppNotes";
import AppViewComponent, { IAppViewProps } from "./AppView";
import ChartInfo from "./ChartInfo";
import ServiceTable from "./ServiceTable";

describe("AppViewComponent", () => {
  // Generates a Yaml file separated by --- containing every object passed.
  const generateYamlManifest = (items: any[]): string => {
    let yamlManifest = "";
    items.forEach(i => {
      yamlManifest += "---\n" + yamlSafeDump(i);
    });
    return yamlManifest;
  };

  const appRelease = hapi.release.Release.create({
    info: hapi.release.Info.create(),
    namespace: "weee",
  });

  const validProps: IAppViewProps = {
    app: appRelease,
    deleteApp: jest.fn(),
    deleteError: undefined,
    error: undefined,
    getApp: jest.fn(),
    namespace: "my-happy-place",
    releaseName: "mr-sunshine",
  };

  context("when app info is null", () => {
    sharedSpecs.loadingSpecs(AppViewComponent, {
      ...validProps,
      app: { ...validProps.app, info: null },
    });
  });

  context("when otherResources is null", () => {
    sharedSpecs.loadingSpecs(AppViewComponent, validProps, { otherResources: null });
  });

  describe("State initialization", () => {
    const resources = {
      configMap: { apiVersion: "v1", kind: "ConfigMap", metadata: { name: "cm-one" } },
      deployment: {
        apiVersion: "extensions/v1beta1",
        kind: "Deployment",
        metadata: { name: "deployment-one" },
      },
      service: { apiVersion: "v1", kind: "Service", metadata: { name: "svc-one" } },
    };

    /*
      The imported manifest contains one deployment, one service, one config map and some bogus manifests.
      We only set websockets for deployment and services
    */
    it("sets a list of web sockets for its deployments and services", () => {
      const manifest = generateYamlManifest([
        resources.deployment,
        resources.service,
        resources.configMap,
      ]);

      const wrapper = shallow(<AppViewComponent {...validProps} />);
      validProps.app.manifest = manifest;
      // setProps again so we trigger componentWillReceiveProps
      wrapper.setProps(validProps);
      const sockets: WebSocket[] = wrapper.state("sockets");

      expect(sockets.length).toEqual(2);
      expect(sockets[0].url).toBe(
        "ws://localhost/api/kube/apis/apps/v1beta1/namespaces/weee/deployments?watch=true&fieldSelector=metadata.name%3Ddeployment-one",
      );
      expect(sockets[1].url).toBe(
        "ws://localhost/api/kube/api/v1/namespaces/weee/services?watch=true&fieldSelector=metadata.name%3Dsvc-one",
      );
    });

    it("stores other k8s resources directly in the state", () => {
      const wrapper = shallow(<AppViewComponent {...validProps} />);
      const manifest = generateYamlManifest([resources.configMap, resources.deployment]);

      validProps.app.manifest = manifest;
      wrapper.setProps(validProps);

      const otherResources: Map<string, IResource> = wrapper.state("otherResources");
      const configMap = otherResources["ConfigMap/cm-one"];
      expect(Object.keys(otherResources).length).toEqual(1);

      // It sets the websocket for the deployment
      const sockets: WebSocket[] = wrapper.state("sockets");
      expect(sockets.length).toEqual(1);

      expect(configMap).toBeDefined();
      expect(configMap.metadata.name).toEqual("cm-one");
    });

    it("does not store empty resources, bogus or without kind attribute", () => {
      const wrapper = shallow(<AppViewComponent {...validProps} />);
      const manifest = generateYamlManifest([
        { apiVersion: "v1", metadata: { name: "cm-one" } },
        {},
        "# This is a comment",
        " ",
      ]);

      validProps.app.manifest = manifest;
      wrapper.setProps(validProps);

      const otherResources: Map<string, IResource> = wrapper.state("otherResources");
      expect(Object.keys(otherResources).length).toBe(0);

      const sockets: WebSocket[] = wrapper.state("sockets");
      expect(sockets.length).toEqual(0);
    });
  });

  describe("renderization", () => {
    it("renders all the elements of an application", () => {
      const wrapper = shallow(<AppViewComponent {...validProps} />);
      expect(wrapper.find(ChartInfo).exists()).toBe(true);
      expect(wrapper.find(DeploymentStatus).exists()).toBe(true);
      expect(wrapper.find(AppControls).exists()).toBe(true);
      expect(wrapper.find(ServiceTable).exists()).toBe(true);
      expect(wrapper.find(AppNotes).exists()).toBe(true);
      expect(wrapper.find(AppDetails).exists()).toBe(true);
    });

    it("renders an error if error prop is set", () => {
      const wrapper = shallow(<AppViewComponent {...validProps} error={new NotFoundError()} />);
      const err = wrapper.find(NotFoundErrorAlert);
      expect(err.exists()).toBe(true);
      expect(err.props()).toMatchObject({
        namespace: "my-happy-place",
        resource: 'Application "mr-sunshine"',
      });
    });

    it("renders a forbidden delete-error if if the deleteError prop is a ForbiddenError", () => {
      const wrapper = shallow(
        <AppViewComponent {...validProps} deleteError={new ForbiddenError()} />,
      );
      const err = wrapper.find(PermissionsErrorAlert);
      expect(err.exists()).toBe(true);
      expect(err.props()).toMatchObject({
        action: 'delete Application "mr-sunshine"',
        namespace: "my-happy-place",
      });
    });

    it("renders a not-found delete-error if the deleteError prop is a NotFound error", () => {
      const wrapper = shallow(
        <AppViewComponent {...validProps} deleteError={new NotFoundError()} />,
      );
      const err = wrapper.find(NotFoundErrorAlert);
      expect(err.exists()).toBe(true);
      expect(err.props()).toMatchObject({
        namespace: "my-happy-place",
        resource: 'Application "mr-sunshine"',
      });
    });
  });
});
