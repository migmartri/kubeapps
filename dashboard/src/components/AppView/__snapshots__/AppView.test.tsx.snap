// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppViewComponent when app info is null loading spinner matches the snapshot 1`] = `
<AppView
  app={
    Object {
      "hooks": Array [],
      "info": null,
      "namespace": "weee",
    }
  }
  deleteApp={[MockFunction]}
  getApp={
    [MockFunction] {
      "calls": Array [
        Array [
          "mr-sunshine",
          "my-happy-place",
        ],
        Array [
          "mr-sunshine",
          "my-happy-place",
        ],
      ],
    }
  }
  namespace="my-happy-place"
  releaseName="mr-sunshine"
>
  <LoadingWrapper
    loaded={false}
  >
    <LoaderSpinner>
      <div
        className="spinner"
      >
        <div
          className="spinner__bounce1"
        />
        <div
          className="spinner__bounce2"
        />
      </div>
    </LoaderSpinner>
  </LoadingWrapper>
</AppView>
`;

exports[`AppViewComponent when otherResources is null loading spinner matches the snapshot 1`] = `
<AppView
  app={
    Object {
      "info": Object {},
      "namespace": "weee",
    }
  }
  deleteApp={[MockFunction]}
  getApp={
    [MockFunction] {
      "calls": Array [
        Array [
          "mr-sunshine",
          "my-happy-place",
        ],
        Array [
          "mr-sunshine",
          "my-happy-place",
        ],
        Array [
          "mr-sunshine",
          "my-happy-place",
        ],
        Array [
          "mr-sunshine",
          "my-happy-place",
        ],
      ],
    }
  }
  namespace="my-happy-place"
  releaseName="mr-sunshine"
>
  <LoadingWrapper
    loaded={false}
  >
    <LoaderSpinner>
      <div
        className="spinner"
      >
        <div
          className="spinner__bounce1"
        />
        <div
          className="spinner__bounce2"
        />
      </div>
    </LoaderSpinner>
  </LoadingWrapper>
</AppView>
`;
